examples/step-55/step-55.cc

```diff
@@ -0,0 +1,1181 @@
+    BlockVector<double> system_rhs;
+
+    MGLevelObject<SparsityPattern>        mg_sparsity_patterns;
+    MGLevelObject<SparseMatrix<double> >  mg_matrices;
+    MGLevelObject<SparseMatrix<double> >  mg_interface_matrices;
+    MGConstrainedDoFs                     mg_constrained_dofs;
+
+    TimerOutput computing_timer;
+  };
+
+
+
+  template <int dim>
+  StokesProblem<dim>::StokesProblem (const unsigned int degree, SolverType::type solver_type)
+    :
+    degree (degree),
+    solver_type (solver_type),
+    triangulation (Triangulation<dim>::maximum_smoothing),
+    fe (FE_Q<dim>(degree+1), dim, // Finite element for whole system
+        FE_Q<dim>(degree), 1),
+    velocity_fe (FE_Q<dim>(degree+1), dim), // Finite element for velocity-only
+    dof_handler (triangulation),
+    velocity_dof_handler (triangulation),
+    computing_timer (std::cout, TimerOutput::summary,
+                     TimerOutput::wall_times)
+  {}
+
+
+
+// @sect4{StokesProblem::setup_dofs}
+
+// This function sets up things differently based on if you want to use ILU or GMG as a preconditioner.
+  template <int dim>
+  void StokesProblem<dim>::setup_dofs ()
+  {
+    TimerOutput::Scope scope(computing_timer, "Setup");
+
+    system_matrix.clear ();
+    pressure_mass_matrix.clear ();
+
+    // We don't need the multigrid dofs for whole problem finite element
+    dof_handler.distribute_dofs(fe);
+
+    // This first creates and array (0,0,1) which means that it first does everything with index 0 and then 1
+    std::vector<unsigned int> block_component (dim+1,0);
+    block_component[dim] = 1;
+
+    // This always knows how to use the dim (start at 0 one)
+    FEValuesExtractors::Vector velocities(0);
+
+    if (solver_type == SolverType::FGMRES_ILU)
+      {
+        TimerOutput::Scope ilu_specific(computing_timer, "(ILU specific)");
+        DoFRenumbering::Cuthill_McKee (dof_handler);
+      }
+
+    DoFRenumbering::component_wise (dof_handler, block_component);
+
+    if (solver_type == SolverType::FGMRES_GMG)
+      {
+        TimerOutput::Scope multigrid_specific(computing_timer, "(Multigrid specific)");
+        TimerOutput::Scope setup_multigrid(computing_timer, "Setup - Multigrid");
+
+        // Distribute only the dofs for velocity finite element
+        velocity_dof_handler.distribute_dofs(velocity_fe);
+
+        // Multigrid only needs the dofs for velocity
+        velocity_dof_handler.distribute_mg_dofs(velocity_fe);
+
+        typename FunctionMap<dim>::type boundary_condition_function_map;
+        BoundaryValuesForVelocity<dim> velocity_boundary_condition;
+        boundary_condition_function_map[0] = &velocity_boundary_condition;
+
+        mg_constrained_dofs.clear();
+        mg_constrained_dofs.initialize(velocity_dof_handler, boundary_condition_function_map);
+        const unsigned int n_levels = triangulation.n_levels();
+
+        mg_interface_matrices.resize(0, n_levels-1);
+        mg_interface_matrices.clear ();
```

Comment:
This second line is confusing because it looks like it clears the `mg_interface_matrices` object. But reading through the documentation of that function, it actually clears the level objects it stores. This will need to be documented.